### PR TITLE
Bug Fix - SQL Server HADR queries for SQL Versions

### LINKED ISSUE #8750: SQLServer plugin: Invalid columns with SQLServerAvailabilityReplicaStates on SQL 2014 and earlier
Columns 'seeding_mode_desc' and 'secondary_lag_seconds' are only available on SQL 2016 and later.
The plugin generates an error on SQL 2014 and earlier when the SQLServerAvailabilityReplicaStates input script is selected. Could a check for version be included to determine which columns to return.

### Relevant telegraf.conf:

```toml
# Read metrics from Microsoft SQL Server
[[inputs.sqlserver]]
  servers = ["Server=xxxxxxxxx;Port=1433"]

  database_type = "SQLServer"
```

### System info:

All Windows versions
SQL 2014 and earlier 
Telegraf version 1.17.0

### Expected behavior:

No error returned regarding invalid column names
### Actual behavior:

2021-01-26T16:21:39Z I! Starting Telegraf 1.17.0
2021-01-26T16:21:39Z I! [inputs.sqlserver] Config: database_type: SQLServer , query_version:0 , azuredb: false
2021-01-26T16:21:39Z E! [inputs.sqlserver] Error in plugin: Script SQLServerAvailabilityReplicaStates failed: mssql: Invalid column name 'seeding_mode_desc'.
2021-01-26T16:21:39Z E! [inputs.sqlserver] Error in plugin: Script SQLServerDatabaseReplicaStates failed: mssql: Invalid column name 'secondary_lag_seconds'.
2021-01-26T16:21:41Z E! [telegraf] Error running agent: input plugins recorded 2 errors

### PR DESCRIPTION
**Pull request by avinash-nigam**
### Required for all PRs:

- [ ] Associated README.md updated.
- [ ] Has appropriate unit tests.

This PR fixes a bug in SqlServer plugin's HADR queries where various SQL Server versions (from 2012 onwards) were not handled, and causes the plugin to report failures for 2012 and 2014 versions.

closes #8750

### CODE REVIEW

**Comment by avinash-nigam** [comment]
@ssoroka @srebhan @Trovalo @denzilribeiro @masree @coquagli @bhsu-ms
Could you please help review this?

**Comment by srebhan** [comment]
Sorry this is beyond my SQL capabilities... :-)

File: plugins/inputs/sqlserver/sqlserverqueries.go [hunk 1]
```diff
@@ -1158,48 +1158,61 @@ IF SERVERPROPERTY('EngineEdition') NOT IN (2,3,4) BEGIN /*NOT IN Standard,Enterp
...
-	from sys.dm_hadr_availability_replica_states AS hars
-	inner join sys.availability_replicas AS ar on hars.replica_id = ar.replica_id
-	inner join sys.availability_groups AS ag on ar.group_id = ag.group_id
-	inner join sys.dm_hadr_availability_group_states AS hags ON hags.group_id = ag.group_id
+DECLARE
+	 @SqlStatement AS nvarchar(max)
+	,@MajorMinorVersion AS int = CAST(PARSENAME(CAST(SERVERPROPERTY('ProductVersion') AS nvarchar),4) AS int)*100 + CAST(PARSENAME(CAST(SERVERPROPERTY('ProductVersion') AS nvarchar),3) AS int)
+	,@Columns AS nvarchar(MAX) = ''
+
+IF @MajorMinorVersion >= 1300 BEGIN
```
Comment (masree) [inline plugins/inputs/sqlserver/sqlserverqueries.go:1166]: Thank you for the fix! Can you also add/update the unit tests for AG queries in sqlserver_test.go?

File: plugins/inputs/sqlserver/sqlserverqueries.go [hunk 2]
```diff
@@ -1209,38 +1222,55 @@ IF SERVERPROPERTY('EngineEdition') NOT IN (2,3,4) BEGIN /*NOT IN Standard,Enterp
...
-		last_commit_time,
-		secondary_lag_seconds
-	from sys.dm_hadr_database_replica_states AS drs
-	inner join sys.availability_replicas AS ar on drs.replica_id = ar.replica_id
+DECLARE
+	 @SqlStatement AS nvarchar(max)
+	,@MajorMinorVersion AS int = CAST(PARSENAME(CAST(SERVERPROPERTY('ProductVersion') AS nvarchar),4) AS int)*100 + CAST(PARSENAME(CAST(SERVERPROPERTY('ProductVersion') AS nvarchar),3) AS int)
+	,@Columns AS nvarchar(MAX) = ''
+
+IF @MajorMinorVersion >= 1200 BEGIN
```
Comment (masree) [inline plugins/inputs/sqlserver/sqlserverqueries.go:1230]: Add a comment above  for readability - 2012 and above version

File: plugins/inputs/sqlserver/sqlserverqueries.go [hunk 3]
```diff
@@ -1209,38 +1222,55 @@ IF SERVERPROPERTY('EngineEdition') NOT IN (2,3,4) BEGIN /*NOT IN Standard,Enterp
...
+	SET @Columns += N'
+	,is_primary_replica'
 END
+
+IF @MajorMinorVersion >= 1300 BEGIN
```
Comment (masree) [inline plugins/inputs/sqlserver/sqlserverqueries.go:1235]: Same as above - 2016 and above versions

**Comment by masree** [comment]
Thank you for the fix!

File: plugins/inputs/sqlserver/sqlserverqueries.go (hunk 1 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserverqueries.go:1168]: Added couple of test cases for AG queries.

File: plugins/inputs/sqlserver/sqlserverqueries.go (hunk 2 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserverqueries.go:1234]: Added a comment at the start of the query describing the query and the version based specification

File: plugins/inputs/sqlserver/sqlserverqueries.go (hunk 3 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserverqueries.go:1239]: Added a comment at the start of the query describing the query and the version based specification

File: plugins/inputs/sqlserver/sqlserverqueries.go (hunk 1 above)
Comment (masree) [reply plugins/inputs/sqlserver/sqlserverqueries.go:1168]: Thanks! Please mark it as resolved!

File: plugins/inputs/sqlserver/sqlserverqueries.go (hunk 3 above)
Comment (masree) [reply plugins/inputs/sqlserver/sqlserverqueries.go:1239]: Please mark all comments as resolved! Thanks!

**Review by masree — APPROVED** [review]

**Review by Trovalo — APPROVED** [review]

**Comment by Trovalo** [comment]
closes #8750

**Comment by avinash-nigam** [comment]
@sjwang90 / @ssoroka Could you please help merge this?

File: plugins/inputs/sqlserver/sqlserverqueries.go [hunk 4]
```diff
@@ -9,11 +9,11 @@ import (
 // Variable @MajorMinorVersion:
 //   - 1000 --> SQL Server 2008
 //   - 1050 --> SQL Server 2008 R2
-//   - 1011 --> SQL Server 2012
```
Comment (helenosheaa) [inline plugins/inputs/sqlserver/sqlserverqueries.go:12]: Just wondering why these are different?

File: plugins/inputs/sqlserver/sqlserverqueries.go [hunk 5]
```diff
@@ -1151,96 +1151,130 @@ FROM (
...
+IF @MajorMinorVersion >= 1300 BEGIN
+	SET @Columns += N'
+	,secondary_lag_seconds'
+END
+
+SET @SqlStatement = N'
+IF SERVERPROPERTY(''IsHadrEnabled'') = 1 BEGIN
+	SELECT
+		''sqlserver_hadr_dbreplica_states'' AS [measurement]
+		,REPLACE(@@SERVERNAME,''\'','':'') AS [sql_instance]
```
Comment (helenosheaa) [inline plugins/inputs/sqlserver/sqlserverqueries.go:1248]: Same question here on the quotes

File: plugins/inputs/sqlserver/sqlserverqueries.go [hunk 6]
```diff
@@ -1151,96 +1151,130 @@ FROM (
...
+	,ag.is_distributed
+	,ar.seeding_mode
+	,ar.seeding_mode_desc'
 END
+
+SET @SqlStatement = N'
+IF SERVERPROPERTY(''IsHadrEnabled'') = 1 BEGIN
+	SELECT
+		''sqlserver_hadr_replica_states'' AS [measurement]
+		,REPLACE(@@SERVERNAME, ''\'', '':'') AS [sql_instance]
```
Comment (helenosheaa) [inline plugins/inputs/sqlserver/sqlserverqueries.go:1180]: I think previously all the double quotes were changed to single but this is going back to double, was this intentional?

**Review by helenosheaa — CHANGES_REQUESTED** [review]
Just minor questions but other than that looks good

File: plugins/inputs/sqlserver/sqlserverqueries.go (hunk 4 above)
Comment (Trovalo) [reply plugins/inputs/sqlserver/sqlserverqueries.go:12]: this corrects a typo

File: plugins/inputs/sqlserver/sqlserverqueries.go (hunk 6 above)
Comment (Trovalo) [reply plugins/inputs/sqlserver/sqlserverqueries.go:1180]: The previous version was just a query being executed as-is. The new one requires dynamic SQL (build and execute a SQL string), in order to put the `'` char inside this SQL string you must use `''`. it's a way to escape that particular char.

**Review by Trovalo — APPROVED** [review]
The code looks good as it is now, I've added the explanation.

**Review by helenosheaa — APPROVED** [review]
Ok thanks for the clarification, happy to merge it now then

**Comment by avinash-nigam** [comment]
Thank you @Trovalo for helping with the clarifications and approval. And thank you @helenosheaa for merging the PR. I'll resolve the open conversation in the above thread.

File: plugins/inputs/sqlserver/sqlserverqueries.go (hunk 5 above)
Comment (avinash-nigam) [reply plugins/inputs/sqlserver/sqlserverqueries.go:1248]: As @Trovalo  mentioned, "The previous version was just a query being executed as-is. The new one requires dynamic SQL (build and execute a SQL string), in order to put the ' char inside this SQL string you must use ''. it's a way to escape that particular char."

**Pull request merged**